File: VideoStreamingBackend/src/application/use-cases/StopStream.ts
/**
 * Use case: Stop a camera stream.
 */
import { StreamSession } from '../../domain/entities/StreamSession';
import { StreamNotFoundError } from '../../domain/errors';
import { StreamManager } from '../../infrastructure/streaming/StreamManager';

export interface StopStreamInput {
  cameraId: string;
}

export class StopStreamUseCase {
  constructor(private readonly streamManager: StreamManager) {}

  execute(input: StopStreamInput): StreamSession {
    const session = this.streamManager.stopStream(input.cameraId);
    if (!session) {
      throw new StreamNotFoundError(input.cameraId);
    }
    return session;
  }
}

File: StreamingBackend/internal/application/usecases/stop_stream.go
package usecases

import (
	"github.com/vigileye/streaming-backend/internal/domain/entities"
	"github.com/vigileye/streaming-backend/internal/infrastructure/streaming"
)

// StopStreamUseCase handles stopping a camera stream.
type StopStreamUseCase struct {
	streamManager *streaming.StreamManager
}

// NewStopStreamUseCase creates a new StopStreamUseCase.
func NewStopStreamUseCase(sm *streaming.StreamManager) *StopStreamUseCase {
	return &StopStreamUseCase{streamManager: sm}
}

// Execute stops a camera stream.
func (uc *StopStreamUseCase) Execute(cameraID string) (*entities.StreamSession, error) {
	return uc.streamManager.StopStream(cameraID)
}
